### ISSUE TITLE
[SimplifyCFG] Missed opportunity to turn a switch into a table lookup

### ISSUE THREAD

**Issue by lattner** [llvm/llvm-project#55667]
The following IR should be changable into a lookup table, we currently leave it to the backend which generates a jump table which is less efficient:

```
%"class.Thing" = type { i8 }
define %"class.Thing"* @test(%"class.Thing"* returned %0, i64 %1) unnamed_addr #0 align 2 {
  %3 = trunc i64 %1 to i8
  switch i8 %3, label %18 [
    i8 66, label %4
    i8 -117, label %5
    i8 -122, label %6
    i8 -115, label %7
    i8 1, label %8
    i8 -119, label %9
    i8 98, label %10
    i8 -121, label %11
    i8 65, label %12
    i8 67, label %13
    i8 99, label %14
    i8 -116, label %15
    i8 -118, label %16
    i8 -120, label %17
  ]

4:                                                ; preds = %2
  br label %18

5:                                                ; preds = %2
  br label %18

6:                                                ; preds = %2
  br label %18

7:                                                ; preds = %2
  br label %18

8:                                                ; preds = %2
  br label %18

9:                                                ; preds = %2
  br label %18

10:                                               ; preds = %2
  br label %18

11:                                               ; preds = %2
  br label %18

12:                                               ; preds = %2
  br label %18

13:                                               ; preds = %2
  br label %18

14:                                               ; preds = %2
  br label %18

15:                                               ; preds = %2
  br label %18

16:                                               ; preds = %2
  br label %18

17:                                               ; preds = %2
  br label %18

18:                                               ; preds = %2, %4, %5, %6, %7, %8, %9, %10, %11, %12, %13, %14, %15, %16, %17
  %19 = phi i8 [ 17, %17 ], [ 16, %16 ], [ 13, %15 ], [ 12, %14 ], [ 11, %13 ], [ 10, %12 ], [ 9, %11 ], [ 8, %10 ], [ 7, %9 ], [ 6, %8 ], [ 4, %7 ], [ 3, %6 ], [ 2, %5 ], [ 1, %4 ], [ 0, %2 ]
  %20 = getelementptr inbounds %"class.Thing", %"class.Thing"* %0, i64 0, i32 0
  store i8 %19, i8* %20, align 1
  ret %"class.Thing"* %0
}

attributes #0 = { nofree norecurse nounwind ssp willreturn writeonly "disable-tail-calls"="false" "frame-pointer"="non-leaf" "less-precise-fpmad"="false" "min-legal-vector-width"="0" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="true" "probe-stack"="__chkstk_darwin" "stack-protector-buffer-size"="8" "target-cpu"="apple-m1" "target-features"="+aes,+crc,+crypto,+dotprod,+fp-armv8,+fp16fml,+fullfp16,+lse,+neon,+ras,+rcpc,+rdm,+sha2,+sha3,+sm4,+v8.5a,+zcm,+zcz" "unsafe-fp-math"="false" "use-soft-float"="false" }

```

Looks like it might just need a minor generalization:

```
$ opt x.ll -simplifycfg -S -o - -debug
Args: opt x.ll -simplifycfg -S -o - -debug 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
Looking to fold  into 
Can't fold, phi node  in  is conflicting with regard to common predecessor 
```

LLC compiles this into 161 lines of .s on aarch64, most of which is a big jump table.

**Comment by alexander-shaposhnikov** [comment]
I've taken a brief look at it, multiple things are going on:

1. without `--switch-to-lookup=true` ConvertSwitchToLookupTable is false => SwitchToLookupTable is not called
2. without a `datalayout` string the vector LegalIntWidths is empty => ShouldBuildLookupTable gives up early
3. if one passes  --switch-to-lookup=true and adds e.g. target datalayout = "e-m:e-i64:64-i128:128-n32:64-S128" the function ShouldBuildLookupTable will almost accomplish its goal except that the last comparison https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L6153 will fail. If it's manually adjusted then finally everything appears to work:

opt ~/cfg.ll -simplifycfg -S -o - --switch-to-lookup=true

```llvm
; ModuleID = '/usr/local/google/home/ashaposhnikov/cfg.ll'
source_filename = "/usr/local/google/home/ashaposhnikov/cfg.ll"
target datalayout = "e-m:e-i64:64-i128:128-n32:64-S128"

%class.Thing = type { i8 }

@switch.table.test = private unnamed_addr constant [222 x i8] c"\03\09\11\07\10\02\0D\04\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\06\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\0A\01\0B\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\08\0C", align 1

; Function Attrs: nofree norecurse nounwind ssp willreturn writeonly
define %class.Thing* @test(%class.Thing* returned %0, i64 %1) unnamed_addr #0 align 2 {
  %3 = trunc i64 %1 to i8
  %switch.tableidx = sub i8 %3, -122
  %4 = icmp ult i8 %switch.tableidx, -34
  br i1 %4, label %switch.lookup, label %5

switch.lookup:                                    ; preds = %2
  %switch.tableidx.zext = zext i8 %switch.tableidx to i9
  %switch.gep = getelementptr inbounds [222 x i8], [222 x i8]* @switch.table.test, i32 0, i9 %switch.tableidx.zext
  %switch.load = load i8, i8* %switch.gep, align 1
  br label %5

5:                                                ; preds = %switch.lookup, %2
  %6 = phi i8 [ 0, %2 ], [ %switch.load, %switch.lookup ]
  %7 = getelementptr inbounds %class.Thing, %class.Thing* %0, i64 0, i32 0
  store i8 %6, i8* %7, align 1
  ret %class.Thing* %0
}

attributes #0 = { nofree norecurse nounwind ssp willreturn writeonly "disable-tail-calls"="false" "frame-pointer"="non-leaf" "less-precise-fpmad"="false" "min-legal-vector-width"="0" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="true" "probe-stack"="__chkstk_darwin" "stack-protector-buffer-size"="8" "target-cpu"="apple-m1" "target-features"="+aes,+crc,+crypto,+dotprod,+fp-armv8,+fp16fml,+fullfp16,+lse,+neon,+ras,+rcpc,+rdm,+sha2,+sha3,+sm4,+v8.5a,+zcm,+zcz" "unsafe-fp-math"="false" "use-soft-float"="false" }
```

**Comment by lattner** [comment]
Ooh, good call re: datalayout.  I should have included that, I'm just using clang on an M1 mac, no tricks up my sleeve! :-).  Out of curiosity, why isn't --switch-to-lookup=true  on by default?

**Comment by alexander-shaposhnikov** [comment]
```shell
opt -O1 -debug-pass-manager ~/cfg.ll -S -o - 2>&1 | grep -i SimplifyCFGPass | wc -l
8
```
and the last pass has this option set to true 
(the comment https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L6629 is relevant)
To make this particular example work the main obstacle is
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L6153 
(we probably need to adjust/relax this comparison or replace it with a better heuristic)

**Comment by zmodem** [comment]
> To make this particular example work the main obstacle is
> https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L6153
> (we probably need to adjust/relax this comparison or replace it with a better heuristic)
> (the comment there also appears to be a bit outdated)

Yes, I think that's what's preventing the lookup table conversion to fire here. It looks like there are 14 cases with a range from -122 to 99. That would give a 6% density for the lookup table which is way below the threshold of 40%.

That 40% was chosen to match the density threshold for jump tables in the backend. That has since been dropped to 10% (still 40% for optsize): https://github.com/llvm/llvm-project/blob/llvmorg-14.0.4/llvm/lib/CodeGen/TargetLoweringBase.cpp#L82
We should probably update simplifycfg to match.

But also, the backend has become smarter and can build jump tables for dense subsections of the switch. I think that's what's happening in this example. Simplifycfg has a FIXME about that: https://github.com/llvm/llvm-project/blob/llvmorg-14.0.4/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L5853 but I'm not really sure we'd want to do that in simplifycfg since it would actually make the cfg more complex.

My pie-in-the-sky vision for switch lowering would be if the lowering in the backend could be rewritten as an IR pass instead, and have lookup table transformation as one of its strategies. That would also benefit other targets which currently have their own switch lowering code. This would be a lot of work though. Also I think when the idea of moving switch lowering to the IR level has come up before, there were concerns about whether it would be able to generate efficient enough code at that level e.g. for jump tables. I'm not sure those concerns are still valid though.

**Comment by lattner** [comment]
Yeah I agree that having it use the same heuristic to split tables that are individually dense is a good way to go.  What is your concern over making the cfg more complex?

given Llvm’s existing architecture, I don’t think we will be able to do this in the backend.  Selection dag at least is single block at a time, and the lowered representation would be very difficult to pattern match and reconstruct a lowered table.

Tldr I think making the simplify-cfg transform smarter and manually keeping it in sync with codegen is the best way to go.

thank you for working on this!

**Comment by zmodem** [comment]
> Yeah I agree that having it use the same heuristic to split tables that are individually dense is a good way to go. What is your concern over making the cfg more complex?

The concern was that if we start peeling off parts of the switch, we're adding more control flow, not simplifying it as one would expect in simplifycfg. But I'm also not sure if this would actually be a problem in practice.

> given Llvm’s existing architecture, I don’t think we will be able to do this in the backend. Selection dag at least is single block at a time, and the lowered representation would be very difficult to pattern match and reconstruct a lowered table.

The idea would be to do this as an IR pass, before selection dag lowering. We already have such an IR switch lowering pass, which I think the AMDGPU target and probably other downstream targets use in llvm/lib/Transforms/Utils/LowerSwitch.cpp, I believe WASM also does some custom switch lowering, and while globalisel reuses some of the selection dag logic it also needs machinery to do the actual lowering. So if we could unify all switch lowering tricks into a single IR pass, I think that would be very beneficial. The question is if it's possible to get good enough code that way, and it would also be a good amount of work.

> Tldr I think making the simplify-cfg transform smarter and manually keeping it in sync with codegen is the best way to go.

Yes, I think that's fine too.

> thank you for working on this!

I didn't touch this code in a long time, and I'm not sure I'll time to look into this any time soon. Happy to do review if someone else wants to jump in though.
